envoy/cilium_network_policy.h

```diff
@@ -0,0 +1,218 @@
+	}
+      }
+
+      bool Matches(const Envoy::Http::HeaderMap& headers) const {
+	// Empty set matches any headers.
+	return Envoy::Router::ConfigUtility::matchHeaders(headers, headers_);
+      }
+
+      std::vector<Envoy::Router::ConfigUtility::HeaderData> headers_; // Allowed if empty.
+    };
+    
+    class PortNetworkPolicyRule : public Logger::Loggable<Logger::Id::config> {
+    public:
+      PortNetworkPolicyRule(const cilium::PortNetworkPolicyRule& rules) {
+	for (const auto& remote: rules.remote_policies()) {
+	  ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicyRule(): Allowing remote {}", remote);
+	  allowed_remotes_.emplace(remote);
+	}
+	for (const auto& http_rule: rules.http_rules().http_rules()) {
+	  http_rules_.emplace_back(http_rule);
+	}
+      }
+
+      bool Matches(uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	// Remote ID must match if we have any.
+	if (allowed_remotes_.size()) {
+	  bool matches = false;
+	  for (const auto& remote: allowed_remotes_) {
+	    if (remote == remote_id) {
+	      matches = true;
+	      break;
+	    }
+	  }
+	  if (!matches) {
+	    return false;
+	  }
+	}
+	// Empty set matches any payload
+	if (http_rules_.size() == 0) {
+	  return true;
+	}	
+	for (const auto& rule: http_rules_) {
+	  if (rule.Matches(headers)) {
+	    return true;
+	  }
+	}
+	return false;
+      }
+
+      std::unordered_set<uint64_t> allowed_remotes_; // Everyone allowed if empty.
+      std::vector<HttpNetworkPolicyRule> http_rules_; // Allowed if empty, but remote is checked first.
+    };
+
+    class PortNetworkPolicyRules : public Logger::Loggable<Logger::Id::config> {
+    public:
+      PortNetworkPolicyRules(const google::protobuf::RepeatedPtrField<cilium::PortNetworkPolicyRule>& rules) {
+	if (rules.size() == 0) {
+	    ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicyRules(): No rules, will allow everything.");
+	}
+	for (const auto& it: rules) {
+	  rules_.emplace_back(PortNetworkPolicyRule(it));
+	}
+      }
+
+      bool Matches(uint64_t remote_id, const Envoy::Http::HeaderMap& headers) const {
+	// Empty set matches any payload from anyone
+	if (rules_.size() == 0) {
+	  return true;
+	}
+	for (const auto& rule: rules_) {
+	  if (rule.Matches(remote_id, headers)) {
+	    return true;
+	  }
+	}
+	return false;
+      }
+
+      std::vector<PortNetworkPolicyRule> rules_; // Allowed if empty.
+    };
+    
+    class PortNetworkPolicy : public Logger::Loggable<Logger::Id::config> {
+    public:
+      PortNetworkPolicy(const google::protobuf::RepeatedPtrField<cilium::PortNetworkPolicy>& rules) {
+	for (const auto& it: rules) {
+	  // TODO: Handle Missing Protocol
+	  if (it.protocol() == envoy::api::v2::core::SocketAddress::TCP) {
+	    // port may be zero, which matches any port.
+	    ENVOY_LOG(trace, "Cilium L7 PortNetworkPolicy(): installing TCP policy for port {}", it.port());
+	    rules_.emplace(it.port(), PortNetworkPolicyRules(it.rules()));
```

Comment:
Would it be useful to add a check that there is no PortNetworkPolicyRules associated with this TCP port yet?
The Cilium agent ensures that, but it may be good to add a check here?

Reply:
I added check for the return value's `second` member which tells if a new element was added. If not, an exception is thrown, which will result in a xDS NACK.